synapse/logging/tracing.py

```diff
@@ -477,102 +434,71 @@ def whitelisted_homeserver(destination: str) -> bool:
 
 # Start spans and scopes
 
-# Could use kwargs but I want these to be explicit
+
 def start_active_span(
-    operation_name: str,
-    child_of: Optional[Union["opentracing.Span", "opentracing.SpanContext"]] = None,
-    references: Optional[List["opentracing.Reference"]] = None,
-    tags: Optional[Dict[str, str]] = None,
-    start_time: Optional[float] = None,
-    ignore_active_span: bool = False,
-    finish_on_close: bool = True,
+    name: str,
     *,
-    tracer: Optional["opentracing.Tracer"] = None,
-) -> "opentracing.Scope":
-    """Starts an active opentracing span.
-
-    Records the start time for the span, and sets it as the "active span" in the
-    scope manager.
-
-    Args:
-        See opentracing.tracer
-    Returns:
-        scope (Scope) or contextlib.nullcontext
-    """
-
-    if opentracing is None:
+    context: Optional["opentelemetry.context.context.Context"] = None,
+    kind: Optional["opentelemetry.trace.SpanKind"] = SpanKind.INTERNAL,
+    attributes: "opentelemetry.util.types.Attributes" = None,
+    links: Optional[Sequence["opentelemetry.trace.Link"]] = None,
+    start_time: Optional[int] = None,
+    record_exception: bool = True,
+    set_status_on_exception: bool = True,
+    end_on_exit: bool = True,
```

Comment:
I'd prefer to just `*args`, `**kargs` this but how do I share the same type hints from `tracer.start_as_current_span`?

Reply:
Regrettably, I don't think there's an easy way to do this at present. I could have sworn I've seen an issue for this on either python/typeshed or python/typing, but I can't seem to find it right now.

We could try to hack something together with a no-op decorator:

```python
P = ParamSpec("P")
R = TypeVar("R")
def inherits_signature(wrapped: Callable[P, R]) -> Callable[[Callable], Callable[P, R]:
    def noop_decorator(wrapper: Callable) -> Callable[P, R]:
        return wrapper  # type: ignore

@inherits_signature(FOO.BAR.BAZ.start_active_span)
def start_active_span(*args: Any, **kwargs: Any) -> Any:
    if opentelemetry is None:
        ...
    tracer = ....
    return tracer.start_active_span(*args, **kwargs)
```

But I'm guessing this won't work unless the `wrapped` function is a free function; `start_active_span` seems to be a method. Could probably work around this if we really tried hard enough...

Reply:
https://stackoverflow.com/questions/71253495/how-to-annotate-the-type-of-arguments-forwarded-to-another-function

https://github.com/python/typeshed/pull/6670 might mean we can use `functools.wraps` for this rather than having to write our own decorator?

Reply:
Thanks @DMRobertson, sorry for the delay, was pushing forward in other areas before looking at this.

Played with this a bit but thinking that I can't reference `@inherits_signature(opentelemetry.trace.Tracer.start_span)` since `opentelemetry` is optionally imported?

What does `functools.wraps` do without that PR? I see we already use it a few places but it's not clear to me what it does it it's not already maintaining the typing.

Reply:
> Played with this a bit but thinking that I can't reference `@inherits_signature(opentelemetry.trace.Tracer.start_span)` since `opentelemetry` is optionally imported?

Urgh, that's not fun. You could maybe try

```python
def start_active_span(...): ...
if TYPE_CHECKING:
    start_active_span = inherits_signature(FOO.BAR.BAZ.start_active_span)
```

but at this point we're probably fighting the typing system rather than working with it; the least bad option is probably going to be manually copying the annotations as you've already done. Brittle and not pretty, but it seems to work.

> What does functools.wraps do without that PR? 

Not a huge amount, tbh. The invocation

```
@functools.wraps(original)
def new_wrapper(...): ...
```

copies [certain pieces metadata](https://docs.python.org/3/library/functools.html?highlight=functools%20wraps#functools.update_wrapper) of `original` onto `new_wrapper` at runtime to make the wrapper as transparent as possible.

Reply:
Thanks again @DMRobertson, will leave it with the copied annotations ⏩